synapse/handlers/auth.py

```diff
@@ -1805,59 +1801,242 @@ def _generate_base_macaroon(self, user_id: str) -> pymacaroons.Macaroon:
+            assert f is not None
+
+            return maybe_awaitable(f(*args, **kwargs))
+
+        return run
+
+    # populate hooks with the implemented methods, wrapped with async_wrapper
+    hooks = {
+        hook: async_wrapper(getattr(provider, hook, None))
+        for hook in password_auth_provider_methods
+    }
+
+    supported_login_types = {}
+    # call get_supported_login_types and add that to the dict
+    g = getattr(provider, "get_supported_login_types", None)
+    if g is not None:
+        # Note the old module style also called get_supported_login_types at loading time
+        # and it is synchronous
+        supported_login_types.update(g())
+
+    auth_checkers = {}
+    # Legacy modules have a check_auth method which expects to be called with one of
+    # the keys returned by get_supported_login_types. New style modules register a
+    # dictionary of login_type->check_auth_method mappings
+    check_auth = async_wrapper(getattr(provider, "check_auth", None))
+    if check_auth is not None:
+        for login_type in supported_login_types:
+            auth_checkers[login_type] = check_auth
+
+    # if it has a "check_password" method then it should handle all auth checks
+    # with login type of LoginType.PASSWORD
+    check_password = async_wrapper(getattr(provider, "check_password", None))
+    if check_password is not None:
+        supported_login_types[LoginType.PASSWORD] = ("password",)
+        auth_checkers[LoginType.PASSWORD] = check_password
+
+    api.register_password_auth_provider_callbacks(
+        hooks, supported_login_types=supported_login_types, auth_checkers=auth_checkers
+    )
+
+
+CHECK_3PID_AUTH_CALLBACK = Callable[
+    [str, str, str], Awaitable[Optional[Tuple[str, Optional[Callable]]]]
+]
+ON_LOGGED_OUT_CALLBACK = Callable[[str, Optional[str], str], Awaitable]
+CHECK_AUTH_CALLBACK = Callable[
+    [str, str, JsonDict],
+    Awaitable[Optional[Tuple[str, Optional[Callable]]]],
+]
+
+
+class PasswordAuthProvider:
+    """
+    A class that the AuthHandler calls when doing authenticating users
+    It allows modules to provide alternative methods for authentication
+    """
+
+    def __init__(self):
+        # lists of callbacks
+        self.check_3pid_auth_callbacks: List[CHECK_3PID_AUTH_CALLBACK] = []
+        self.on_logged_out_callbacks: List[ON_LOGGED_OUT_CALLBACK] = []
+
+        # Mapping from login type to login parameters
+        self._supported_login_types: Dict[str, Iterable[str]] = {}
+
+        # Mapping from login type to auth checker callbacks
+        self.auth_checker_callbacks: Dict[str, List[CHECK_AUTH_CALLBACK]] = {}
+
+    def register_password_auth_provider_callbacks(
+        self,
+        check_3pid_auth: Optional[CHECK_3PID_AUTH_CALLBACK] = None,
+        on_logged_out: Optional[ON_LOGGED_OUT_CALLBACK] = None,
+        supported_login_types: Optional[Dict[str, Iterable[str]]] = None,
+        auth_checkers: Optional[Dict[str, CHECK_AUTH_CALLBACK]] = None,
```

Comment:
Yes I think it would make more sense to register them as `Dict[Tuple[str, Iterable[str]], CHECK_AUTH_CALLBACK]` 

However I'm not really sure about splitting them up into two different register methods: for example in which one do you put your callback for `check_3pid_auth`? It is an *auth checker* but it's meant to handle just 3pid authentication and so is the same format...

Reply:
> However I'm not really sure about splitting them up into two different register methods: for example in which one do you put your callback for `check_3pid_auth`? It is an auth checker but it's meant to handle just 3pid authentication and so is the same format...

That's a good point. Maybe we should change how the call stack around auth checkers work to allow 3pid auth to be checked in the same callback format as other checkers, but that's definitely out of scope for this PR.